File: src/vector.c
#include "vector.h"
#define ARRAY_CLEARED(name, type, size) \
    type name[size] = { 0 }; \
    memset(&name[0], 0, sizeof(name));

static char*
generate_vector(char* Type) {
    char* Name = calloc(100, sizeof(char));
    strcat(Name, "vector_");
    strcat(Name, Type);

    // Make sure that the Type of vector container
    // that is getting generated does not already exist.
    if (VecTypes.Type == NULL) {
        VecTypes.Type = calloc(strlen(Type), sizeof(char));
        strcpy(VecTypes.Type, Type);
    } else {
        if (not_in_list(&VecTypes, Type)) {
            add_type_to_list(&VecTypes, Type);
        } else {
            return Name;
        }
    }

    // Guard to protect against redefinition
    ADD_TO_VEC("#ifndef VECTOR_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC("_\n");
    ADD_TO_VEC("#define VECTOR_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC("_\n");

    // Generate the vec struct
    // Typedef struct {
    //     Type* Items;
    //     int TotSize;
    //     int CurSize;
    // } vector_Type;
    ADD_TO_VEC("typedef struct {\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC(Type);
    ADD_TO_VEC("* Items;\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("int TotSize;\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("int CurSize;\n");
    ADD_TO_VEC("} ");
    ADD_TO_VEC(Name);
    ADD_TO_VEC(";\n");

    // Generate the function calls
    //
    //
    // void Vector_'Type'_init(Vector_'Type' *Vec) {
    //      Vec->Items = malloc(100 * sizeof('Type'));
    //      Vec->TotSize = 100;
    //      Vec->CurSize = 0;
    // }
    ADD_TO_VEC("void ");
    ADD_TO_VEC(Name);
    ADD_TO_VEC("_init(");
    ADD_TO_VEC(Name);
    ADD_TO_VEC(" *Vec) {\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("Vec->Items = malloc(100 * sizeof(");
    ADD_TO_VEC(Type);
    ADD_TO_VEC("));\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("Vec->TotSize = 100;\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("Vec->CurSize = 0;\n");
    ADD_TO_VEC("}\n");

    // void vector_'Type'_expand(vector_'Type' *Vec) {
    //     Vec->TotSize = Vec->TotSize  * 2;
    //     Vec->Items = realloc(Vec->Items, sizeof('Type') * Vec->TotSize);
    // }
    ADD_TO_VEC("void vector_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC("_expand(vector_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC(" *Vec) {\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("Vec->TotSize = Vec->TotSize * 2;\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("Vec->Items = realloc(Vec->Items, sizeof(");
    ADD_TO_VEC(Type);
    ADD_TO_VEC(") * Vec->TotSize);\n");
    ADD_TO_VEC("}\n");

    // void Vector_'Type'_push(Vector_'Type' *Vec, 'Type' Item) {
    //      if (Vec->TotSize == Vec->CurSize) {
    //      }
    //      Vec->Items[Vec->CurSize++] = Item;
    //  }
    ADD_TO_VEC("void ");
    ADD_TO_VEC(Name);
    ADD_TO_VEC("_push(");
    ADD_TO_VEC(Name);
    ADD_TO_VEC(" *Vec, ");
    ADD_TO_VEC(Type);
    ADD_TO_VEC(" Item) {\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("if (Vec->TotSize == Vec->CurSize) {\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("vector_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC("_expand(Vec);\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("}\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("Vec->Items[Vec->CurSize++] = Item;\n");
    ADD_TO_VEC("}\n");

    //void vector_'Type'_insert(vector_'Type' *Vec, 'Type' Pos, 'Type' Item) {
    //    for ('Type' i = Vec->CurSize + 1; i > Pos - 1; i--) {
    //        Vec->Items[i+1] = Vec->Items[i]; 
    //    }
    //    Vec->Items[Pos] = Item;
    //    Vec->CurSize++;
    //}
    ADD_TO_VEC("void vector_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC("_insert(vector_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC(" *Vec, int Pos, ");
    ADD_TO_VEC(Type);
    ADD_TO_VEC(" Item) {\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("if (Vec->TotSize == Vec->CurSize) {\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("vector_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC("_expand(Vec);\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("}\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("for (int i = Vec->CurSize + 1; i > Pos - 1; i--) {\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("Vec->Items[i+1] = Vec->Items[i];\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("}\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("Vec->Items[Pos] = Item;\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("Vec->CurSize++;\n");
    ADD_TO_VEC("}\n");

    //  inline 'Type'* vector_'Type'_at(vector_'Type' Vec, int Pos) {
    //      return &Vec.Items[Pos]
    //  }
    ADD_TO_VEC("static inline ");
    ADD_TO_VEC(Type);
    ADD_TO_VEC("* vector_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC("_at(vector_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC(" Vec, int Pos) {\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("return &Vec.Items[Pos];\n");
    ADD_TO_VEC("}\n");

    // inline 'Type'* vector_'Type'_front(vector_'Type' Vec) {
    //      return &Vec.Items[0];
    // }
    ADD_TO_VEC("static inline ");
    ADD_TO_VEC(Type);
    ADD_TO_VEC("* vector_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC("_front(vector_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC(" Vec) {\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("return &Vec.Items[0];\n");
    ADD_TO_VEC("}\n");

    // static inline void vector_'Type'_free(vector_'Type' *Vec) {
    //      free(Vec->Items);
    //      Vec->Items = 0;
    //      Vec->CurSize = 0;
    //      Vec->TotSize = 0;
    // }
    ADD_TO_VEC("static inline void vector_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC("_free(vector_");
    ADD_TO_VEC(Type);
    ADD_TO_VEC(" *Vec) {\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("free(Vec->Items);\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("Vec->Items = 0;\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("Vec->CurSize = 0;\n");
    ADD_TO_VEC(TAB);
    ADD_TO_VEC("Vec->TotSize = 0;\n");
    ADD_TO_VEC("}\n");

    // End the redefinition guard
    ADD_TO_VEC("#endif\n");

    return Name;
}

static bool
parse_vector_var(char* VarName) {
    get_next_token();
    if (CurTok == '[') {
        ADD_TO_NORMAL("*vector_");
        ADD_TO_NORMAL(get_var_type(&VecVariables, VarName));
        ADD_TO_NORMAL("_at(");
        ADD_TO_NORMAL(VarName);
        ADD_TO_NORMAL(", ");
        
        get_next_token();
		if (CurTok == tok_constant || CurTok == tok_identifier) {
		    ADD_TO_NORMAL(IdentifierStr);
		} else if (CurTok > 0) {
            NormalOutput[NormalIndex++] = (char) CurTok;
        } else {
        	return log_error("Expected number after '['", CurrentLine);
        }
        ADD_TO_NORMAL(")");
		
    } else {
        ARRAY_CLEARED(Error, char, 100);
        strcat(Error, "Unknown symbol after ");
        strcat(Error, VarName);
        strcat(Error, ".");
        return log_error(Error, CurrentLine);
    }
    get_next_token(); // eat ']'
    
    return true;
}

static bool
parse_vector_function() {
    get_next_token();

    if (strcmp(IdentifierStr, "push") == 0) {
        // vector_push(var, value);
        get_next_token(); // Eat 'push'

        get_next_token(); // Eat '('

        func_call_args Args = get_func_args();

        // Expect the first argument to be VarName
        char* VarName = Args.Arg;

        char* VarType = get_var_type(&VecVariables, VarName);
        if (!VarType) {
            return false;
        }

        ARRAY_CLEARED(FuncName, char, 100);
        strcat(FuncName, "vector_");
        strcat(FuncName, VarType);
        strcat(FuncName, "_push");

        ADD_TO_NORMAL(FuncName);
        ADD_TO_NORMAL("(&");
        ADD_TO_NORMAL(VarName);
        ADD_TO_NORMAL(", ");
        ADD_TO_NORMAL(Args.Next->Arg);

        output_remaining_func_line();
        free_func_call_args(Args);
    }
    if (strcmp(IdentifierStr, "at") == 0) {
        // vector_at(var, index);
        get_next_token(); // Eat 'at'
        
        get_next_token(); // Eat '('
        
        func_call_args Args = get_func_args();

        // Expect the first argument to be VarName
        char* VarName = Args.Arg;

        char* VarType = get_var_type(&VecVariables, VarName);
        ARRAY_CLEARED(FuncName, char, 100);
        strcat(FuncName, "*vector_");
        strcat(FuncName, VarType);
        strcat(FuncName, "_at");

        ADD_TO_NORMAL(FuncName);
        ADD_TO_NORMAL("(");
        ADD_TO_NORMAL(VarName);
        ADD_TO_NORMAL(", ");
        ADD_TO_NORMAL(Args.Next->Arg);

        output_remaining_func_line();
        free_func_call_args(Args);
    }
    if (strcmp(IdentifierStr, "front") == 0) {
        // vector_front(var)
        get_next_token(); // Eat 'front'
        get_next_token(); // Eat '('

        func_call_args Args = get_func_args();

        // Expect the first argument to be VarName
        char* VarName = Args.Arg;

        char* VarType = get_var_type(&VecVariables, VarName);
        ARRAY_CLEARED(FuncName, char, 100);
        strcat(FuncName, "*vector_");
        strcat(FuncName, VarType);
        strcat(FuncName, "_front");

        ADD_TO_NORMAL(FuncName);
        ADD_TO_NORMAL("(");
        ADD_TO_NORMAL(VarName);

        output_remaining_func_line();
        free_func_call_args(Args);
    }

    if (strcmp(IdentifierStr, "insert") == 0) {
        // vector_insert(var, pos, item)
        get_next_token(); // Eat 'insert'
        get_next_token(); // Eat '('

        func_call_args Args = get_func_args();

        // Expect the first argument to be VarName
        char* VarName = Args.Arg;

        char* VarType = get_var_type(&VecVariables, VarName);
        ARRAY_CLEARED(FuncName, char, 100);
        strcat(FuncName, "vector_");
        strcat(FuncName, VarType);
        strcat(FuncName, "_insert(&");

        ADD_TO_NORMAL(FuncName);

        func_call_args* Iterator = &Args;
        while (Iterator->Next != NULL) {
            ADD_TO_NORMAL(Iterator->Arg); 
            ADD_TO_NORMAL(", ");

            Iterator = Iterator->Next;
        }
        ADD_TO_NORMAL(Iterator->Arg);

        output_remaining_func_line();
        free_func_call_args(Args);
    }

    if (strcmp(IdentifierStr, "free") == 0) {
        // vector_free(var)
        get_next_token(); // Eat 'free'
        get_next_token(); // Eat '('

        func_call_args Args = get_func_args();

        // Expect the first argument to be VarName
        char* VarName = Args.Arg;

        char* VarType = get_var_type(&VecVariables, VarName);
        ARRAY_CLEARED(FuncName, char, 100);
        strcat(FuncName, "vector_");
        strcat(FuncName, VarType);
        strcat(FuncName, "_free");

        ADD_TO_NORMAL(FuncName);
        ADD_TO_NORMAL("(&");
        ADD_TO_NORMAL(VarName);

        output_remaining_func_line();
        free_func_call_args(Args);
    }
    return true;
}

// Caller Requirements: free returned value
static char*
get_vector_file_name() {
    char* VecOutName = (char *)calloc(100, sizeof(char));
    strcat(VecOutName, "__vector");
    ARRAY_CLEARED(Num, char, 10);
    sprintf(Num,"%d",SourceFileCount);
    strcat(VecOutName, Num);
    strcat(VecOutName, ".h");
    return VecOutName;
}

static bool
parse_vector() {
	if (CurVecFile < SourceFileCount) {
        ARRAY_CLEARED(IncludeStr, char, 100);

        strcat(IncludeStr, "#include \"");
        char* FileName = get_vector_file_name();
        strcat(IncludeStr, FileName);
        free(FileName);
        strcat(IncludeStr, "\"\n");
		CurVecFile++;
		ADD_TO_GLOBAL(IncludeStr);

        // Need this for malloc and realloc
        ADD_TO_VEC("#include <stdlib.h>\n");
	}
	

	// Parse argument inside of '< >'
    get_next_token(); // Eat 'vector'

    if (CurTok == '_')
        return parse_vector_function();

    if (CurTok != '<')
        return log_error("Expected '<' in vector declaration", CurrentLine);
    get_next_token(); // Eat '<'

    char Type[100] = { 0 };
    strcpy(Type, IdentifierStr);
    char* VecName = generate_vector(Type);

    get_next_token();
    if (CurTok != '>')
        return log_error("Expected '>' in vector declaration", CurrentLine);

    char VarName[100] = { 0 };

    // Look for the var name
    while (CurTok != tok_identifier) {
        // If you hit the semicolon before finding a var name
        if (CurTok == ';') {
            return log_error("Expected variable name in vector declaration", CurrentLine);
        }

        get_next_token();
    }
    while (CurTok != ';') {
        if (CurTok == tok_identifier) {
            strcat(VarName, IdentifierStr);
        } else {
            int Len = strlen(VarName);
            VarName[Len] = CurTok;
        }
        get_next_token();
    }
    get_next_token(); // eat the newline


    if (VecVariables.Name == NULL) {
        VecVariables.Name = calloc(strlen(VarName), sizeof(char));
        strcpy(VecVariables.Name, VarName);

        VecVariables.Type = calloc(strlen(Type), sizeof(char));
        strcpy(VecVariables.Type, Type);
    } else {
        add_var_to_list(&VecVariables, Type, VarName);
    }

    // Change the variable name declaration
    // vector_'Type' name;
    ADD_TO_NORMAL(VecName);
    ADD_TO_NORMAL(" ");
    free(VecName);
    ADD_TO_NORMAL(VarName);
    ADD_TO_NORMAL(";\n");

    // Insert a call to vector_'Type'_init
    add_tabs_to_normal_output();
    ADD_TO_NORMAL("vector_");
    ADD_TO_NORMAL(Type);
    ADD_TO_NORMAL("_init(&");
    ADD_TO_NORMAL(VarName);
    ADD_TO_NORMAL(");\n");

    return true;
}

static void
empty_vec_types() {
    empty_type_list(&VecTypes);
}

File: src/vector.py
from logging import log_error
from output import Output
from c_token import CToken, Tok

from collections import deque
from typing import Dict, List

def get_func_arg(tokens: deque) -> str:
    arg = ""
    while 1:
        token = tokens[0]
        if token.string == ",":
            return arg
        elif token.string.isspace():
            return arg
        elif token.string == ")":
            return arg
        elif token.val == Tok.identifier:
            arg += token.string
        else:
            arg += token.string
        tokens.popleft()


def get_func_args(tokens: deque) -> List:
    token = tokens[0]
    args = []
    while token.string != ")":
        while token.string == "," or token.string.isspace():
            tokens.popleft()
            token = tokens[0]
        args.append(get_func_arg(tokens))
        token = tokens[0]
    return args

class Vector:
    output: Output
    variables: Dict[str, str] = dict()
    init_list: List[str] = []

    # should be given output.vector_out
    def __init__(self, output: Output):
        self.output = output
        # need the standard lib for malloc/realloc
        output.vector_out += "#include <stdlib.h>\n"

    def purge_variables(self, variables: List[str]) -> None:
        for var in variables:
            if var in self.variables:
                del self.variables[var]


    def parse(self, tokens: deque) -> str:
        normal_out: str = self.output.normal_out
        token: CToken
        var_name: str = ""
        vec_name: str = ""
        vec_type: str = ""

        tokens.popleft() # eat 'vector'
        token = tokens[0]
        if token.string == "_":
            self.parse_function(tokens)
            return var_name
        if token.string != "<":
            log_error(token, "Expected '<' in vector declaration")
        tokens.popleft() # eat '<'

        vec_type = tokens[0].string
        vec_name = self.generate_definition(vec_type)
        print(vec_name)

        tokens.popleft() # eat '>'
        token = tokens[0]
        if token.string != ">":
            log_error(token, "Expected '>' in vector declaration")

        # make sure you don't hit the end of statment before finding a var name
        while token.val != Tok.identifier:
            if token.val == Tok.semicolon:
                log_error(token, "Expected variable name in vector declaration")
            tokens.popleft()
            token = tokens[0]

        # get whole var name
        while token.val != Tok.semicolon and token.string != ")":
            var_name += token.string
            tokens.popleft()
            token = tokens[0]

        # The variable should be discarded when leaving a function
        self.variables[var_name] = vec_type

        while (token.val != Tok.semicolon) and (token.string != ")"):
            token = tokens.popleft()

        if (token.val == Tok.semicolon):
            tokens.popleft()
            tokens.popleft()
            # vector_'type' name;
            normal_out += vec_name + " " + var_name + ";\n"
        elif (token.string == ")"):
            normal_out += vec_name + " " + var_name


        self.output.normal_out = normal_out
        return var_name
    def parse_variable(self, tokens: deque, var_name: str) -> None:
        if tokens[0].string == "[":
            tokens.popleft()
            normal_out = self.output.normal_out
            normal_out += "*vector_" + self.get_var_type(var_name, tokens[0])
            normal_out += "_at(" + var_name + ", "

            token = tokens[0]
            if token.val == Tok.constant or token.val == Tok.identifier or token.val == Tok.char:
                normal_out += token.string
            else:
                log_error(token, "Expected a number or variable after '['")
            tokens.popleft()
            tokens.popleft()

            normal_out += ")"
            self.output.normal_out = normal_out
        else:
            self.output.normal_out += var_name

        return

    def get_var_type(self, var_name: str, token: CToken) -> str:
        var_type = ""
        if var_name in self.variables:
            var_type = self.variables[var_name]
        else: 
            log_error(token, "Variable '" + var_name + "' does not exist.")
        return var_type

    def parse_function(self, tokens: deque) -> None:
        normal_out = self.output.normal_out
        tokens.popleft() # "eat '_'
        token = tokens[0]
        var_type = ""

        if token.string == "push":
            tokens.popleft() # "eat 'push'
            tokens.popleft() # "eat '('
            args = get_func_args(tokens)
            if len(args) != 2:
                log_error(tokens[0], "Invalid number of arguments in call to 'vector_push'")

            var_name = args[0] # expect first arg to be name
            var_type = self.get_var_type(var_name, tokens[0])

            normal_out += "vector_" + var_type + "_push"
            normal_out += "(&" + var_name + ", " + args[1]
        if token.string == "at":
            tokens.popleft()
            tokens.popleft()
            args = get_func_args(tokens)
            if len(args) != 2:
                log_error(tokens[0], "Invalid number of arguments in call to 'vector_at'")
            var_name = args[0]
            var_type = self.get_var_type(var_name, tokens[0])
            normal_out += "*vector_" + var_type
            normal_out += "_at(" + var_name + ", " + args[1]
        if token.string == "front":
            tokens.popleft() # Eat 'front'
            tokens.popleft() # Eat '('
            args = get_func_args(tokens)
            if len(args) != 1:
                log_error(tokens[0], "Invalid number of arguments in call to 'vector_front'")
            var_name = args[0]
            var_type = self.get_var_type(var_name, tokens[0])
            normal_out += "*vector_" + var_type + "_front(" + var_name
        if token.string == "insert":
            tokens.popleft() # Eat 'insert'
            tokens.popleft() # Eat '('
            args = get_func_args(tokens)
            if len(args) != 3:
                log_error(tokens[0], "Invalid number of arguments in call to 'vector_insert'")
            var_name = args[0]
            var_type = self.get_var_type(var_name, tokens[0])
            normal_out += "vector_" + var_type + "_insert(&"
            normal_out += args[0] + ", " + args[1] + ", " + args[2]
        if token.string == "free":
            tokens.popleft() # Eat 'free'
            tokens.popleft() # Eat '('
            args = get_func_args(tokens)
            if len(args) != 1:
                log_error(tokens[0], "Invalid number of arguments in call to 'vector_free'")
            var_name = args[0]
            var_type = self.get_var_type(var_name, tokens[0])
            normal_out += "vector_" + var_type + "_free"
            normal_out += "(&" + var_name
        if token.string == "init":
            tokens.popleft() # Eat 'init'
            tokens.popleft() # Eat '('
            args = get_func_args(tokens)
            if len(args) != 1:
                log_error(tokens[0], "Invalid number of arguments in call to 'vector_init'")
            var_name = args[0]
            var_type = self.get_var_type(var_name, tokens[0])
            normal_out += "vector_" + var_type + "_init"
            normal_out += "(&" + var_name



        token = tokens[0]
        while token.val != Tok.semicolon:
            normal_out += token.string
            tokens.popleft()
            token = tokens[0]
        normal_out += ";"
        tokens.popleft()
        self.output.normal_out = normal_out




    def generate_definition(self, vec_type: str) -> str:
        tab = "    "
        output = self.output.vector_out
        name = "vector_" + vec_type

        output += "#ifndef VECTOR_" + vec_type + "_\n"
        output += "#define VECTOR_" + vec_type + "_\n"

        # generate the vec struct
        # Typedef struct {
        #     Type* Items;
        #     int TotSize;
        #     int CurSize;
        # } vector_Type;
        output += "typedef struct {\n"
        output += tab + vec_type + "* items;\n"
        output += tab + "int tot_size;\n"
        output += tab + "int cur_size;\n"
        output += "} vector_" + vec_type + ";\n"


        # void Vector_'Type'_init(Vector_'Type' *Vec) {
        #      Vec->Items = malloc(100 * sizeof('Type'));
        #      Vec->TotSize = 100;
        #      Vec->CurSize = 0;
        # }
        output += "void " + name + "_init(" + name + " *vec) {\n"
        output += tab + "vec->items = "
        output += "(" + vec_type + " *) "
        output +="malloc(100 * sizeof(" + vec_type + "));\n"
        output += tab + "vec->tot_size = 100;\n"
        output += tab + "vec->cur_size = 0;\n"
        output += "}\n"

        # void vector_'Type'_expand(vector_'Type' *Vec) {
        #     Vec->TotSize = Vec->TotSize  * 2;
        #     Vec->Items = realloc(Vec->Items, sizeof('Type') * Vec->TotSize);
        # }
        output += "void vector_" + vec_type + "_expand(vector_" + vec_type + " *vec) {\n"
        output += tab + "vec->tot_size = vec->tot_size * 2;\n"
        output += tab + "vec->items = (" + vec_type + " *) "
        output += "realloc(vec->items, sizeof(" + vec_type + ") * vec->tot_size);\n"
        output += "}\n"

        # void Vector_'Type'_push(Vector_'Type' *Vec, 'Type' Item) {
        #      if (Vec->TotSize == Vec->CurSize) {
        #      }
        #      Vec->Items[Vec->CurSize++] = Item;
        #  }
        output += "void " + name + "_push(" + name + " *vec, " + vec_type + " item) {\n"
        output += tab + "if (vec->tot_size == vec->cur_size) {\n"
        output += tab + tab + "vector_" + vec_type + "_expand(vec);\n"
        output += tab + "}\n"
        output += tab + "vec->items[vec->cur_size++] = item;\n"
        output += "}\n"

        # void vector_'Type'_insert(vector_'Type' *Vec, 'Type' Pos, 'Type' Item) {
        #    for ('Type' i = Vec->CurSize + 1; i > Pos - 1; i--) {
        #        Vec->Items[i+1] = Vec->Items[i]; 
        #    }
        #    Vec->Items[Pos] = Item;
        #    Vec->CurSize++;
        # }
        output += "void vector_" + vec_type + "_insert(vector_" + vec_type + " *vec, int pos, " + vec_type + " item) {\n"
        output += tab + "for (int i = vec->cur_size + 1; i > pos - 1; i--) {\n"
        output += tab + tab + "vec->items[i+1] = vec->items[i];\n"
        output += tab + "}\n"
        output += tab + "vec->items[pos] = item;\n"
        output += tab + "vec->cur_size++;\n"
        output += "}\n"

        #  inline 'Type'* vector_'Type'_at(vector_'Type' Vec, int Pos) {
        #      return &Vec.Items[Pos]
        #  }
        output += "static inline " + vec_type + "* vector_" + vec_type +"_at(vector_" +vec_type + " vec, int pos) {\n"
        output += tab + "return &vec.items[pos];\n"
        output += "}\n"

        # inline 'Type'* vector_'Type'_front(vector_'Type' vec) {
        #       return &vec->items[0]
        # }
        output += "static inline " + vec_type + "* vector_" + vec_type
        output += "_front(vector_" + vec_type + " vec) {\n"
        output += tab + "return &vec.items[0];\n"
        output += "}\n"

        
        # static inline void vector_'Type'_free(vector_'Type' *Vec) {
        #      free(Vec->Items);
        #      Vec->Items = 0;
        #      Vec->CurSize = 0;
        #      Vec->TotSize = 0;
        # }
        output += "static inline void vector_" + vec_type + "_free(vector_" + vec_type + " *vec) {\n"
        output += tab + "free(vec->items);\n"
        output += tab + "vec->items = 0;\n"
        output += tab + "vec->cur_size = 0;\n"
        output += tab + "vec->tot_size = 0;\n"
        output += "}\n"

        output += "#endif // VECTOR_" + vec_type + "_\n"

        self.output.vector_out = output
        return name
